File: Core/Services/BranchAnalyticsEngine.swift
import Foundation

// MARK: - Branch analytics
//
// Stage 1.11 — Branch Performance Overview. Pure functions that turn the raw
// repository data (branches + performance scores + roster counts) into the
// executive analytics the dashboard renders. No framework imports — this is
// `Core/Services`, sibling to `ScoreEngine`.
//
// Composite, grade and the six metric scores are real averages of seeded
// `PerformanceScore` data. Growth %, the 12-week attendance trend and the
// coaching radar are demo-derived: deterministic functions of the branch's
// real scores + a stable per-branch seed, so the dashboard is populated and
// reproducible without a separate analytics table.

/// The six headline performance dimensions shown as metric rings.
public enum BranchMetricKind: String, CaseIterable, Sendable, Hashable {
    case competition, technique, fitness, attendance, progress, wellness

    public var labelKey: String { "branch.metric.\(rawValue)" }
}

/// The six coaching-effectiveness radar axes.
public enum BranchRadarAxis: String, CaseIterable, Sendable, Hashable {
    case athleteImprovement, technicalQuality, attendanceImpact
    case retentionRate, promotionSuccess, sessionQuality

    public var labelKey: String { "branch.radar.\(rawValue)" }
}

/// Direction-of-travel chip.
public enum BranchTrend: String, Sendable, Hashable {
    case improving, stable, declining

    public var labelKey: String { "branch.trend.\(rawValue)" }
}

public struct BranchMetricScore: Identifiable, Sendable, Hashable {
    public let kind: BranchMetricKind
    public var score: Double          // 0–100
    public var id: String { kind.rawValue }

    public init(kind: BranchMetricKind, score: Double) {
        self.kind = kind
        self.score = score
    }
}

public struct BranchRadarScore: Identifiable, Sendable, Hashable {
    public let axis: BranchRadarAxis
    public var score: Double          // 0–100
    public var id: String { axis.rawValue }

    public init(axis: BranchRadarAxis, score: Double) {
        self.axis = axis
        self.score = score
    }
}

/// Everything a branch ranking card + the bottom charts need.
public struct BranchAnalytics: Identifiable, Sendable, Hashable {
    public let id: EntityID
    public var branch: Branch
    public var rank: Int
    public var nationalRank: Int
    public var athleteCount: Int
    public var coachCount: Int
    public var sessionsPerWeek: Int
    public var compositeScore: Double          // 0–100
    public var grade: LetterGrade
    public var metrics: [BranchMetricScore]    // six, in `BranchMetricKind.allCases` order
    public var radar: [BranchRadarScore]       // six, in `BranchRadarAxis.allCases` order
    public var growthPct: Double               // 30-day, may be negative
    public var trend: BranchTrend
    public var attendanceTrend: [Double]        // 12 weekly values, 0–100

    public init(
        id: EntityID, branch: Branch, rank: Int, nationalRank: Int,
        athleteCount: Int, coachCount: Int, sessionsPerWeek: Int,
        compositeScore: Double, grade: LetterGrade,
        metrics: [BranchMetricScore], radar: [BranchRadarScore],
        growthPct: Double, trend: BranchTrend, attendanceTrend: [Double]
    ) {
        self.id = id
        self.branch = branch
        self.rank = rank
        self.nationalRank = nationalRank
        self.athleteCount = athleteCount
        self.coachCount = coachCount
        self.sessionsPerWeek = sessionsPerWeek
        self.compositeScore = compositeScore
        self.grade = grade
        self.metrics = metrics
        self.radar = radar
        self.growthPct = growthPct
        self.trend = trend
        self.attendanceTrend = attendanceTrend
    }

    public func metric(_ kind: BranchMetricKind) -> Double {
        metrics.first { $0.kind == kind }?.score ?? 0
    }
}

/// A single executive insight. The view formats the localized sentence from
/// these structured fields.
public struct BranchInsight: Identifiable, Sendable, Hashable {
    public enum Kind: String, Sendable, Hashable {
        case improvedAttendance, topCompetition, needsAttention, overallGrowth
    }
    public let id: EntityID
    public var kind: Kind
    public var branchName: String
    public var value: Double

    public init(id: EntityID = UUID(), kind: Kind, branchName: String, value: Double) {
        self.id = id
        self.kind = kind
        self.branchName = branchName
        self.value = value
    }
}

// MARK: - Engine

public enum BranchAnalyticsEngine {

    /// Raw per-branch inputs gathered from the repository.
    public struct Input: Sendable {
        public var branch: Branch
        public var scores: [PerformanceScore]
        public var athleteCount: Int
        public var coachCount: Int
        public var sessionsPerWeek: Int

        public init(branch: Branch, scores: [PerformanceScore],
                    athleteCount: Int, coachCount: Int, sessionsPerWeek: Int) {
            self.branch = branch
            self.scores = scores
            self.athleteCount = athleteCount
            self.coachCount = coachCount
            self.sessionsPerWeek = sessionsPerWeek
        }
    }

    /// Builds the full analytics set, sorted best-composite first and ranked.
    public static func analyze(_ inputs: [Input]) -> [BranchAnalytics] {
        let unranked = inputs.map { build($0) }
        let sorted = unranked.sorted { $0.compositeScore > $1.compositeScore }
        return sorted.enumerated().map { idx, a in
            var ranked = a
            ranked.rank = idx + 1
            ranked.nationalRank = idx + 1
            return ranked
        }
    }

    private static func build(_ input: Input) -> BranchAnalytics {
        let scores = input.scores
        func avg(_ kp: KeyPath<PerformanceScore, Double>) -> Double {
            guard !scores.isEmpty else { return 0 }
            return scores.reduce(0.0) { $0 + $1[keyPath: kp] } / Double(scores.count)
        }
        let metricPairs: [(BranchMetricKind, Double)] = [
            (.competition, avg(\.competition)),
            (.technique,   avg(\.technical)),
            (.fitness,     avg(\.physical)),
            (.attendance,  avg(\.adherence)),
            (.progress,    avg(\.beltProgression)),
            (.wellness,    avg(\.wellness)),
        ]
        let metrics = metricPairs.map { BranchMetricScore(kind: $0.0, score: $0.1) }
        let composite = scores.isEmpty ? 0 : ScoreEngine.branchComposite(scores)

        // Deterministic 30-day growth: a stable per-branch seed nudged toward
        // the branch's standing — strong branches trend up, weak ones down.
        let s = seed(input.branch.name)
        let growth = (((composite - 70) / 5) + (s * 18 - 7)).rounded(toPlaces: 1)
        let trend: BranchTrend = growth > 3 ? .improving : (growth < -1 ? .declining : .stable)

        // 12-week attendance trend around the attendance metric.
        let base = metricPairs.first { $0.0 == .attendance }?.1 ?? 80
        let trendSeries = (0..<12).map { week -> Double in
            let progress = Double(week) / 11.0
            let drift = (growth / 100) * 14 * (progress - 0.5)
            let wobble = (seed("\(input.branch.name)-w\(week)") - 0.5) * 6
            return min(100, max(40, base + drift + wobble))
        }

        // Coaching radar — six axes derived from the metrics + a small seed.
        let radar = BranchRadarAxis.allCases.map { axis -> BranchRadarScore in
            BranchRadarScore(axis: axis, score: radarScore(axis, metrics: metricPairs,
                                                           composite: composite, branch: input.branch.name))
        }

        return BranchAnalytics(
            id: input.branch.id, branch: input.branch, rank: 0, nationalRank: 0,
            athleteCount: input.athleteCount, coachCount: input.coachCount,
            sessionsPerWeek: input.sessionsPerWeek,
            compositeScore: composite, grade: LetterGrade.from(score: composite),
            metrics: metrics, radar: radar,
            growthPct: growth, trend: trend, attendanceTrend: trendSeries
        )
    }

    private static func radarScore(_ axis: BranchRadarAxis,
                                   metrics: [(BranchMetricKind, Double)],
                                   composite: Double, branch: String) -> Double {
        func m(_ k: BranchMetricKind) -> Double { metrics.first { $0.0 == k }?.1 ?? composite }
        let basis: Double
        switch axis {
        case .athleteImprovement: basis = m(.progress)
        case .technicalQuality:   basis = m(.technique)
        case .attendanceImpact:   basis = m(.attendance)
        case .retentionRate:      basis = (m(.wellness) + m(.attendance)) / 2
        case .promotionSuccess:   basis = (m(.progress) + m(.competition)) / 2
        case .sessionQuality:     basis = (m(.technique) + m(.fitness)) / 2
        }
        let jitter = (seed("\(branch)-\(axis.rawValue)") - 0.5) * 10
        return min(100, max(35, basis + jitter))
    }

    /// Top executive insights derived from the ranked analytics.
    public static func insights(_ analytics: [BranchAnalytics], overallGrowth: Double) -> [BranchInsight] {
        guard !analytics.isEmpty else { return [] }
        var out: [BranchInsight] = []

        if let improver = analytics.filter({ $0.growthPct > 0 }).max(by: { $0.growthPct < $1.growthPct }) {
            out.append(BranchInsight(kind: .improvedAttendance,
                                     branchName: improver.branch.name,
                                     value: improver.growthPct))
        }
        if let topComp = analytics.max(by: { $0.metric(.competition) < $1.metric(.competition) }) {
            out.append(BranchInsight(kind: .topCompetition,
                                     branchName: topComp.branch.name,
                                     value: topComp.metric(.competition)))
        }
        if let weakest = analytics.min(by: { $0.compositeScore < $1.compositeScore }) {
            out.append(BranchInsight(kind: .needsAttention,
                                     branchName: weakest.branch.name,
                                     value: weakest.compositeScore))
        }
        out.append(BranchInsight(kind: .overallGrowth, branchName: "", value: overallGrowth))
        return out
    }
}

// MARK: - Helpers

private extension BranchAnalyticsEngine {
    /// Stable 0–1 value derived from a string — a tiny FNV-1a hash so demo
    /// analytics are reproducible across launches (unlike `hashValue`).
    static func seed(_ string: String) -> Double {
        var hash: UInt64 = 0xcbf29ce484222325
        for byte in string.utf8 {
            hash ^= UInt64(byte)
            hash = hash &* 0x100000001b3
        }
        return Double(hash % 1000) / 1000.0
    }
}

private extension Double {
    func rounded(toPlaces places: Int) -> Double {
        let factor = pow(10.0, Double(places))
        return (self * factor).rounded() / factor
    }
}

File: flutter_app/lib/core/services/branch_analytics_engine.dart
import 'dart:math' as math;

import '../models/branch.dart';
import '../models/entity_id.dart';
import '../models/performance_score.dart';
import 'score_engine.dart';

/// 1:1 port of Core/Services/BranchAnalyticsEngine.swift.
///
/// Pure functions that turn raw repository data (branches + performance scores
/// + roster counts) into the executive analytics the dashboard renders.
///
/// Composite, grade and the six metric scores are real averages of seeded
/// [PerformanceScore] data. Growth %, the 12-week attendance trend and the
/// coaching radar are demo-derived: deterministic functions of the branch's
/// real scores + a stable per-branch seed.

// ---------------------------------------------------------------------------
// Supporting types

/// The six headline performance dimensions shown as metric rings.
enum BranchMetricKind {
  competition,
  technique,
  fitness,
  attendance,
  progress,
  wellness;

  String get labelKey => 'branch.metric.$name';
}

/// The six coaching-effectiveness radar axes.
enum BranchRadarAxis {
  athleteImprovement,
  technicalQuality,
  attendanceImpact,
  retentionRate,
  promotionSuccess,
  sessionQuality;

  String get labelKey => 'branch.radar.$name';
}

/// Direction-of-travel chip.
enum BranchTrend {
  improving,
  stable,
  declining;

  String get labelKey => 'branch.trend.$name';
}

class BranchMetricScore {
  final BranchMetricKind kind;
  final double score; // 0–100

  const BranchMetricScore({required this.kind, required this.score});

  String get id => kind.name;
}

class BranchRadarScore {
  final BranchRadarAxis axis;
  final double score; // 0–100

  const BranchRadarScore({required this.axis, required this.score});

  String get id => axis.name;
}

/// Everything a branch ranking card + the bottom charts need.
class BranchAnalytics {
  final EntityID id;
  final Branch branch;
  final int rank;
  final int nationalRank;
  final int athleteCount;
  final int coachCount;
  final int sessionsPerWeek;
  final double compositeScore; // 0–100
  final LetterGrade grade;
  final List<BranchMetricScore> metrics; // six, in BranchMetricKind order
  final List<BranchRadarScore> radar; // six, in BranchRadarAxis order
  final double growthPct; // 30-day, may be negative
  final BranchTrend trend;
  final List<double> attendanceTrend; // 12 weekly values, 0–100

  const BranchAnalytics({
    required this.id,
    required this.branch,
    required this.rank,
    required this.nationalRank,
    required this.athleteCount,
    required this.coachCount,
    required this.sessionsPerWeek,
    required this.compositeScore,
    required this.grade,
    required this.metrics,
    required this.radar,
    required this.growthPct,
    required this.trend,
    required this.attendanceTrend,
  });

  BranchAnalytics copyWith({int? rank, int? nationalRank}) => BranchAnalytics(
    id: id,
    branch: branch,
    rank: rank ?? this.rank,
    nationalRank: nationalRank ?? this.nationalRank,
    athleteCount: athleteCount,
    coachCount: coachCount,
    sessionsPerWeek: sessionsPerWeek,
    compositeScore: compositeScore,
    grade: grade,
    metrics: metrics,
    radar: radar,
    growthPct: growthPct,
    trend: trend,
    attendanceTrend: attendanceTrend,
  );

  double metric(BranchMetricKind kind) => metrics
      .firstWhere(
        (m) => m.kind == kind,
        orElse: () => BranchMetricScore(kind: kind, score: 0),
      )
      .score;
}

/// A single executive insight.
class BranchInsight {
  final EntityID id;
  final BranchInsightKind kind;
  final String branchName;
  final double value;

  BranchInsight({
    EntityID? id,
    required this.kind,
    required this.branchName,
    required this.value,
  }) : id = id ?? newEntityId();
}

enum BranchInsightKind {
  improvedAttendance,
  topCompetition,
  needsAttention,
  overallGrowth,
}

// ---------------------------------------------------------------------------
// Engine

class BranchAnalyticsEngine {
  BranchAnalyticsEngine._();

  /// Raw per-branch inputs gathered from the repository.
  // ignore: library_private_types_in_public_api
  static const _fnvOffset = 0xcbf29ce484222325;
  static const _fnvPrime = 0x100000001b3;

  /// Builds the full analytics set, sorted best-composite first and ranked.
  static List<BranchAnalytics> analyze(List<BranchAnalyticsInput> inputs) {
    final unranked = inputs.map(_build).toList();
    unranked.sort((a, b) => b.compositeScore.compareTo(a.compositeScore));
    return unranked
        .asMap()
        .entries
        .map((e) => e.value.copyWith(rank: e.key + 1, nationalRank: e.key + 1))
        .toList();
  }

  static BranchAnalytics _build(BranchAnalyticsInput input) {
    final scores = input.scores;

    double avg(double Function(PerformanceScore) fn) {
      if (scores.isEmpty) return 0;
      return scores.fold<double>(0, (acc, s) => acc + fn(s)) / scores.length;
    }

    final metricPairs = [
      (BranchMetricKind.competition, avg((s) => s.competition)),
      (BranchMetricKind.technique, avg((s) => s.technical)),
      (BranchMetricKind.fitness, avg((s) => s.physical)),
      (BranchMetricKind.attendance, avg((s) => s.adherence)),
      (BranchMetricKind.progress, avg((s) => s.beltProgression)),
      (BranchMetricKind.wellness, avg((s) => s.wellness)),
    ];
    final metrics = metricPairs
        .map((p) => BranchMetricScore(kind: p.$1, score: p.$2))
        .toList();
    final composite = scores.isEmpty
        ? 0.0
        : ScoreEngine.branchComposite(scores);

    // Deterministic 30-day growth: a stable per-branch seed nudged toward
    // the branch's standing — strong branches trend up, weak ones down.
    final s = _seed(input.branch.name);
    final growth = _roundToPlaces(((composite - 70) / 5) + (s * 18 - 7), 1);
    final trend = growth > 3
        ? BranchTrend.improving
        : (growth < -1 ? BranchTrend.declining : BranchTrend.stable);

    // 12-week attendance trend around the attendance metric.
    final base = metricPairs
        .firstWhere(
          (p) => p.$1 == BranchMetricKind.attendance,
          orElse: () => (BranchMetricKind.attendance, 80.0),
        )
        .$2;
    final trendSeries = List<double>.generate(12, (week) {
      final progress = week / 11.0;
      final drift = (growth / 100) * 14 * (progress - 0.5);
      final wobble = (_seed('${input.branch.name}-w$week') - 0.5) * 6;
      return (base + drift + wobble).clamp(40.0, 100.0);
    });

    // Coaching radar — six axes derived from the metrics + a small seed.
    final radar = BranchRadarAxis.values
        .map(
          (axis) => BranchRadarScore(
            axis: axis,
            score: _radarScore(axis, metricPairs, composite, input.branch.name),
          ),
        )
        .toList();

    return BranchAnalytics(
      id: input.branch.id,
      branch: input.branch,
      rank: 0,
      nationalRank: 0,
      athleteCount: input.athleteCount,
      coachCount: input.coachCount,
      sessionsPerWeek: input.sessionsPerWeek,
      compositeScore: composite,
      grade: LetterGrade.fromScore(composite),
      metrics: metrics,
      radar: radar,
      growthPct: growth,
      trend: trend,
      attendanceTrend: trendSeries,
    );
  }

  static double _radarScore(
    BranchRadarAxis axis,
    List<(BranchMetricKind, double)> metricPairs,
    double composite,
    String branch,
  ) {
    double m(BranchMetricKind k) => metricPairs
        .firstWhere((p) => p.$1 == k, orElse: () => (k, composite))
        .$2;

    final double basis;
    switch (axis) {
      case BranchRadarAxis.athleteImprovement:
        basis = m(BranchMetricKind.progress);
        break;
      case BranchRadarAxis.technicalQuality:
        basis = m(BranchMetricKind.technique);
        break;
      case BranchRadarAxis.attendanceImpact:
        basis = m(BranchMetricKind.attendance);
        break;
      case BranchRadarAxis.retentionRate:
        basis =
            (m(BranchMetricKind.wellness) + m(BranchMetricKind.attendance)) / 2;
        break;
      case BranchRadarAxis.promotionSuccess:
        basis =
            (m(BranchMetricKind.progress) + m(BranchMetricKind.competition)) /
            2;
        break;
      case BranchRadarAxis.sessionQuality:
        basis =
            (m(BranchMetricKind.technique) + m(BranchMetricKind.fitness)) / 2;
        break;
    }
    final jitter = (_seed('$branch-${axis.name}') - 0.5) * 10;
    return (basis + jitter).clamp(35.0, 100.0);
  }

  /// Top executive insights derived from the ranked analytics.
  static List<BranchInsight> insights(
    List<BranchAnalytics> analytics,
    double overallGrowth,
  ) {
    if (analytics.isEmpty) return [];
    final out = <BranchInsight>[];

    final improvers = analytics.where((a) => a.growthPct > 0).toList();
    if (improvers.isNotEmpty) {
      final improver = improvers.reduce(
        (a, b) => a.growthPct > b.growthPct ? a : b,
      );
      out.add(
        BranchInsight(
          kind: BranchInsightKind.improvedAttendance,
          branchName: improver.branch.name,
          value: improver.growthPct,
        ),
      );
    }

    final topComp = analytics.reduce(
      (a, b) =>
          a.metric(BranchMetricKind.competition) >
              b.metric(BranchMetricKind.competition)
          ? a
          : b,
    );
    out.add(
      BranchInsight(
        kind: BranchInsightKind.topCompetition,
        branchName: topComp.branch.name,
        value: topComp.metric(BranchMetricKind.competition),
      ),
    );

    final weakest = analytics.reduce(
      (a, b) => a.compositeScore < b.compositeScore ? a : b,
    );
    out.add(
      BranchInsight(
        kind: BranchInsightKind.needsAttention,
        branchName: weakest.branch.name,
        value: weakest.compositeScore,
      ),
    );

    out.add(
      BranchInsight(
        kind: BranchInsightKind.overallGrowth,
        branchName: '',
        value: overallGrowth,
      ),
    );
    return out;
  }

  // ---------------------------------------------------------------------------
  // Helpers

  /// Stable 0–1 value derived from a string — FNV-1a hash so demo analytics
  /// are reproducible across launches (unlike hashCode).
  /// Matches the Swift implementation exactly (same constants, 64-bit wrapping).
  static double _seed(String string) {
    var hash = _fnvOffset;
    for (final byte in string.codeUnits) {
      hash ^= byte;
      // Dart ints are 63-bit on VM / 53-bit precision in JS; mask to 64-bit
      // unsigned using the same wrapping semantics as Swift's `&*`.
      hash = _u64Mul(hash, _fnvPrime);
    }
    return (hash % 1000) / 1000.0;
  }

  /// 64-bit unsigned wrapping multiply (approximates Swift's `&*` operator
  /// on UInt64). Dart's integers are arbitrary precision so we cap manually.
  static int _u64Mul(int a, int b) {
    // Split into 32-bit halves to avoid overflow in JS (where int is 53-bit).
    final aLo = a & 0xFFFFFFFF;
    final aHi = (a >> 32) & 0xFFFFFFFF;
    final bLo = b & 0xFFFFFFFF;
    final bHi = (b >> 32) & 0xFFFFFFFF;

    final lo = aLo * bLo;
    final mid1 = aLo * bHi;
    final mid2 = aHi * bLo;
    // We only keep the lower 64 bits (hi contributions wrap away).
    final result = lo + ((mid1 + mid2) << 32);
    // Mask to 64 bits.  Dart's int is 64-bit on native; on JS limit to 2^53.
    return result & 0xFFFFFFFFFFFFFFFF;
  }

  static double _roundToPlaces(double value, int places) {
    final factor = math.pow(10.0, places);
    return (value * factor).roundToDouble() / factor;
  }
}

/// Raw per-branch inputs gathered from the repository.
class BranchAnalyticsInput {
  final Branch branch;
  final List<PerformanceScore> scores;
  final int athleteCount;
  final int coachCount;
  final int sessionsPerWeek;

  const BranchAnalyticsInput({
    required this.branch,
    required this.scores,
    required this.athleteCount,
    required this.coachCount,
    required this.sessionsPerWeek,
  });
}
